### DISCUSSION TITLE
Interacting with Minecraft using TCP

### DISCUSSION THREAD

**Discussion by edoCsItahW** [PrismarineJS/mineflayer#3390] [Q&A] [answered]
Minelayer is indeed a commendable library, and I am learning about the TCP protocol through it. 

I tried to interact with Minecraft's Spigot server by imitating Minelayer's way and content of sending packets, but I encountered an issue.

The process of sending the package is as follows:

## handshake pack
```
{
    protocol_vision: 765,
    ip: 127.0.0.1,
    port: 25565,
    next_state: 2
}
```
I actually sent a hexadecimal packet like this:
`\x10   \x00    \xfd     \x05\t127.0.0.1c  \xdd    \x02`
 ` len    id    vision           ip         port     status`

## login start pack
```
{
    user_name: Player,
    UUID: eb175f15-7757-37b6-977b-24c4f51a1b15
}
```
Same as above
`\x18\x00\x06Player\xa0\x1e8C\xe5!9\x98\x95\x8a\xf4Y\x80\x0eM\x11`

Now, there's a problem here
I send a package

## login acknowledged 
```
{
    void
}
```
like this
`\x01  \x03`
` len   id`
right?

But the server gave me feedback on an error: 
```
Player (/127.0.0.1:51232) lost connection: Internal Exception: io.netty.handler.codec.DecoderException: Badly compressed packet - size of 3 is below server threshold of 256
```

So can any experienced individuals tell me what problems have arisen.

thanks

**Comment by rom1504** [comment]
You can read https://wiki.vg/Protocol

Minecraft expects a specific protocol, you cannot simply take a guess and have it working.

**Reply by edoCsItahW** [reply]
# Thank you for your answer

But in fact, I have referred to this wiki and have not found the crux of the problem. I have tried to request help on Minecraft forums in our country, but there are not many people discussing this topic, which is why I am here

> Minecraft Version 1.20.4
> Protocol Version 765

## I referred to this paragraph
![屏幕截图 2024-06-04 093052](https://github.com/PrismarineJS/mineflayer/assets/139570866/ec3a3ca7-f43e-49db-b430-1857eef1153a)

## And referred to the debugging information of Mineflayer
![屏幕截图 2024-06-04 094629](https://github.com/PrismarineJS/mineflayer/assets/139570866/9966930c-189f-4d92-96d4-eb56418d0414)

## The data I actually sent
![屏幕截图 2024-06-04 100109](https://github.com/PrismarineJS/mineflayer/assets/139570866/43e8a23e-2db0-430d-a3a7-638c27483a5f)

For safety reasons, I uploaded these images, but personally, I think it's more likely that some steps were missing

Finally, thank you for sharing with Minecraft. It's really enjoyable to have robots do various things, but unfortunately, using this library requires me to possess the Nodejs skills that I am still learning.

**Comment by edoCsItahW** [comment]
# New progress

I noticed that the server sent a packet: `\x03\x03\x80\x02`,and the data packet with ID 3 is set for compression,so my login confirmation packet with ID 3 is considered to be set for compression?

## Other

I have discovered some related websites again, although there are only a small amount of related content among them

* [Minecraft Modern](https://wiki.vg/Category:Minecraft_Modern)
* [How to Write a Client](https://wiki.vg/How_to_Write_a_Client)
* [Protocol FAQ](https://wiki.vg/Protocol_FAQ)

**Reply by edoCsItahW** [reply]
I have three options now. 

One is to compress so that I need to learn zlib compression technology. 

The other is to use negative values to disable compression (but does VarInt type support negative values?). 

The third is to ignore the package, but this doesn't seem to work because the login confirmation package ID is also 3.

**Comment by edoCsItahW** [comment]
# New progress

Continuing from what we said last time

## first time
* The first time I tried to follow the wiki's statement of disabling negative thresholds, I sent a packet: `\x06\x03\xff\xff\xff\xff\x0f`

* The server returned an error as a result: `Badly compressed packet - size of 3 is below server threshold of 256`

That may be the reason why an ID is not needed

## second time
* So I sent a package for the second time: `\x05\xff\xff\xff\xff\xff\x0f`

* The response from the server is: `Badly compressed packet - size of -1 is below server threshold of 256`

But this did not achieve the goal of disabling compression

## other
Afterwards, I noticed that

<img width="451" alt="屏幕截图 2024-06-27 140403" src="https://github.com/PrismarineJS/mineflayer/assets/139570866/d1a2c1e0-9e55-4739-8f74-90f974845b9b">

* So I sent a package with 0: `\x07\x00\x03\xff\xff\xff\xff\xff\x0f`

* This time I was probably successful, and the server returned me a long string: `\xcc\x01\x84\x02x\x9cM\x8d1N\x031\x10E\x1d...`

I searched for the package with ID 1 that the server may have returned, but I still couldn't interpret it. The package is probably compressed

## finally
But this sending was not the most successful until I sent: `\x02\x00\x03`

The content sent by the server this time is normal and meets expectations

~~To be honest, the wiki has written quite extensively, but it seems that there are still some examples missing, and many things still need me to guess~~

**Comment by edoCsItahW** [comment]
# Non optimal answer

## resolvent
* According to this paragraph:
<img width="451" alt="屏幕截图 2024-06-27 140403" src="https://github.com/PrismarineJS/mineflayer/assets/139570866/1271c7dc-02bc-49db-9dc0-d506f372d7ba">

**For packets with a length less than the threshold, we only need to add 0 to their 'length' and 'ID', for example: `\x01\x03`becomes `\x02\x00\x03`**

So we just need to send out the `login confirmation` package in this form

## Non optimal reasons
* The best way is actually to set a negative threshold to disable compression, but I haven't found the correct method

## New issues
After sending the `Login Confirmation` package, I sent the `Knowledge Finish Configuration` package.

The wiki said that the ID of this package was 3, but I imitated Mineplayer and sent the package with ID 2, and the result was successful. 

The server sent me a lot of data, but most of it was compressed. 

However, the biggest success was that I saw my role on the server. 

Although this only lasted for a short time, I couldn't find the `Clientbound Keep Alive` package sent to me by the server, and of course, I couldn't respond to it, which led to me being considered a timeout and kicked out of the server.

**Reply by Myitian** [reply]
- Client -> Server packet with ID=3 is not "**Set Compression**". It is "**Login Acknowledged**".
- Client -> Server packet with ID=2 is "**Serverbound Plugin Message**" in Configuration phase (because it is sent after Login Acknowledged).

**Comment by Myitian** [comment, marked as answer]
> The best way is actually to set a negative threshold to disable compression

No, you cannot set the compression threshold on the client side. The SetCompression packet is used for the server to inform the client to set the compression threshold, not for the client to inform the server.

You can read the decompiled Minecraft code to learn more. For example, the SetCompression packet is named `LoginCompressionS2CPacket` in the [yarn mappings](https://github.com/FabricMC/yarn) (Mapping table commonly used by Fabric loader), and you can find its usage: When the client receives the packet, the client will call the `ClientConnection.setCompressionThreshold` method, which sets the compression threshold (or disables compression).

So you should add decompression to your program. Decompressing zlib data in Python is not difficult; just [`import zlib`](https://docs.python.org/library/zlib.html).

However, if you need to log in to your Microsoft account on your Minecraft client, you will also need to encrypt/decrypt data packets.

**Reply by Myitian** [reply]
I comes from ZhiHu :)
https://www.zhihu.com/question/658799238

**Reply by Myitian** [reply]
PS: The wiki.vg was shut down on November 30, 2024. You can now find its content on [Minecraft Wiki:Projects/wiki.vg merge](https://minecraft.wiki/w/Minecraft_Wiki:Projects/wiki.vg_merge/).
